Concatenating empty with empty leads to failed assertion on Dotty

Given 
```scala
import strawman.collection.immutable
```
the code 
```scala
LazyList.empty ++ LazyList.empty
```
leads to a compiler assertion error when compiling with Dotty.

The problem can be avoided by adding a type ascription on any of the ```empty``` calls or by using the ```apply``` method instead of ```empty```, e.g. neither
```scala
LazyList.empty[Nothing] ++ LazyList.empty
``` 
nor
```scala
LazyList.empty ++ LazyList()
``` 
will trigger the assertion error.

Only ```Seq```, ```List``` and ```LazyList``` are affected (i.e. inheritors of ```Seq``` and ```LinearSeq```).

The stack trace of the assertion failure is:
```
java.lang.AssertionError: assertion failed
	at scala.Predef$.assert(Predef.scala:156)
	at dotty.tools.dotc.core.ConstraintHandling.approximation(ConstraintHandling.scala:225)
	at dotty.tools.dotc.core.ConstraintHandling.instanceType(ConstraintHandling.scala:261)
	at dotty.tools.dotc.core.Types$TypeVar.instantiate(Types.scala:3149)
	at dotty.tools.dotc.typer.Inferencing$.op$$anonfun$28(Inferencing.scala:255)
	at scala.compat.java8.JProcedure2.apply(JProcedure2.java:18)
	at scala.compat.java8.JProcedure2.apply(JProcedure2.java:10)
	at dotty.tools.dotc.util.SimpleMap$Map3.foreachBinding(SimpleMap.scala:103)
	at dotty.tools.dotc.typer.Inferencing$.op$80(Inferencing.scala:256)
	at dotty.tools.dotc.typer.Inferencing$.interpolate$2(Inferencing.scala:221)
	at dotty.tools.dotc.typer.Inferencing$.interpolateUndetVars(Inferencing.scala:264)
	at dotty.tools.dotc.typer.Typer.op$91(Typer.scala:1735)
	at dotty.tools.dotc.typer.Typer.op$89(Typer.scala:1733)
	at dotty.tools.dotc.typer.Typer.adapt(Typer.scala:1732)
	at dotty.tools.dotc.typer.Typer.op$111(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.ProtoTypes$FunProto.typedArgs$$anonfun$1$$anonfun$1(ProtoTypes.scala:229)
	at dotty.tools.dotc.typer.ProtoTypes$FunProto.cacheTypedArg(ProtoTypes.scala:215)
	at dotty.tools.dotc.typer.ProtoTypes$FunProto.typedArgs$$anonfun$1(ProtoTypes.scala:229)
	at dotty.tools.dotc.core.Decorators$ListDecorator$.loop$5(Decorators.scala:62)
	at dotty.tools.dotc.core.Decorators$ListDecorator$.mapconserve$extension(Decorators.scala:78)
	at dotty.tools.dotc.typer.ProtoTypes$FunProto.typedArgs(ProtoTypes.scala:229)
	at dotty.tools.dotc.typer.Applications.op$58(Applications.scala:1338)
	at dotty.tools.dotc.typer.Applications.resolveOverloaded(Applications.scala:1256)
	at dotty.tools.dotc.typer.Applications.op$62(Applications.scala:1242)
	at dotty.tools.dotc.typer.Applications.resolveOverloaded(Applications.scala:1195)
	at dotty.tools.dotc.typer.Typer.adaptOverloaded$1(Typer.scala:1795)
	at dotty.tools.dotc.typer.Typer.adaptInterpolated(Typer.scala:2080)
	at dotty.tools.dotc.typer.Typer.op$91(Typer.scala:1737)
	at dotty.tools.dotc.typer.Typer.op$89(Typer.scala:1733)
	at dotty.tools.dotc.typer.Typer.adapt(Typer.scala:1732)
	at dotty.tools.dotc.typer.Typer.op$111(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1641)
	at dotty.tools.dotc.typer.Applications.op$65(Applications.scala:643)
	at dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:641)
	at dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:741)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1526)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1572)
	at dotty.tools.dotc.typer.Typer.op$111(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1641)
	at dotty.tools.dotc.typer.Typer.op$113(Typer.scala:608)
	at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:600)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1533)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1572)
	at dotty.tools.dotc.typer.Typer.op$111(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1641)
	at dotty.tools.dotc.typer.Typer.op$98(Typer.scala:1237)
	at dotty.tools.dotc.typer.Typer.typedDefDef(Typer.scala:1219)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1514)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1571)
	at dotty.tools.dotc.typer.Typer.op$111(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Typer.traverse$4(Typer.scala:1609)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer.op$124(Typer.scala:1319)
	at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:1267)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1517)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1571)
	at dotty.tools.dotc.typer.Typer.op$111(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Typer.traverse$4(Typer.scala:1609)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer.op$120(Typer.scala:1432)
	at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:1419)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1556)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1572)
	at dotty.tools.dotc.typer.Typer.op$111(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Typer.traverse$4(Typer.scala:1620)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer.op$120(Typer.scala:1432)
	at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:1419)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1556)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1572)
	at dotty.tools.dotc.typer.Typer.op$111(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Typer.traverse$4(Typer.scala:1620)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer.op$120(Typer.scala:1432)
	at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:1419)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1556)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1572)
	at dotty.tools.dotc.typer.Typer.op$111(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1641)
	at dotty.tools.dotc.typer.FrontEnd.typeCheck$$anonfun$1(FrontEnd.scala:64)
	at scala.compat.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:32)
	at dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:68)
	at dotty.tools.dotc.typer.FrontEnd.runOn$$anonfun$7(FrontEnd.scala:93)
	at scala.compat.java8.JProcedure1.apply(JProcedure1.java:18)
	at scala.compat.java8.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at dotty.tools.dotc.typer.FrontEnd.runOn(FrontEnd.scala:93)
	at dotty.tools.dotc.Run.$anonfun$$anonfun$13(Run.scala:82)
	at scala.compat.java8.JProcedure1.apply(JProcedure1.java:18)
	at scala.compat.java8.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:90)
	at scala.compat.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.util.Stats$.monitorHeartBeat(Stats.scala:76)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:95)
	at dotty.tools.dotc.Run.compileSources(Run.scala:64)
	at dotty.tools.dotc.Run.compile(Run.scala:48)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:26)
	at dotty.tools.dotc.Driver.process(Driver.scala:124)
	at xsbt.CachedCompilerImpl.run(CompilerInterface.scala:63)
	at xsbt.CachedCompilerImpl.run(CompilerInterface.scala:53)
	at xsbt.CompilerInterface.run(CompilerInterface.scala:37)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at sbt.compiler.AnalyzingCompiler.call(AnalyzingCompiler.scala:107)
	at sbt.compiler.AnalyzingCompiler.compile(AnalyzingCompiler.scala:53)
	at sbt.compiler.AnalyzingCompiler.compile(AnalyzingCompiler.scala:47)
	at sbt.compiler.MixedAnalyzingCompiler$$anonfun$compileScala$1$1.apply$mcV$sp(MixedAnalyzingCompiler.scala:50)
	at sbt.compiler.MixedAnalyzingCompiler$$anonfun$compileScala$1$1.apply(MixedAnalyzingCompiler.scala:50)
	at sbt.compiler.MixedAnalyzingCompiler$$anonfun$compileScala$1$1.apply(MixedAnalyzingCompiler.scala:50)
	at sbt.compiler.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:74)
	at sbt.compiler.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:49)
	at sbt.compiler.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:64)
	at sbt.compiler.IC$$anonfun$compileInternal$1.apply(IncrementalCompiler.scala:160)
	at sbt.compiler.IC$$anonfun$compileInternal$1.apply(IncrementalCompiler.scala:160)
	at sbt.inc.IncrementalCompile$$anonfun$doCompile$1.apply(Compile.scala:66)
	at sbt.inc.IncrementalCompile$$anonfun$doCompile$1.apply(Compile.scala:64)
	at sbt.inc.IncrementalCommon.cycle(IncrementalCommon.scala:32)
	at sbt.inc.Incremental$$anonfun$1.apply(Incremental.scala:72)
	at sbt.inc.Incremental$$anonfun$1.apply(Incremental.scala:71)
	at sbt.inc.Incremental$.manageClassfiles(Incremental.scala:99)
	at sbt.inc.Incremental$.compile(Incremental.scala:71)
	at sbt.inc.IncrementalCompile$.apply(Compile.scala:54)
	at sbt.compiler.IC$.compileInternal(IncrementalCompiler.scala:160)
	at sbt.compiler.IC$.incrementalCompile(IncrementalCompiler.scala:138)
	at sbt.Compiler$.compile(Compiler.scala:155)
	at sbt.Compiler$.compile(Compiler.scala:141)
	at sbt.Defaults$.sbt$Defaults$$compileIncrementalTaskImpl(Defaults.scala:886)
	at sbt.Defaults$$anonfun$compileIncrementalTask$1.apply(Defaults.scala:877)
	at sbt.Defaults$$anonfun$compileIncrementalTask$1.apply(Defaults.scala:875)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

--- a/src/main/scala/strawman/collection/immutable/LazyList.scala
+++ b/src/main/scala/strawman/collection/immutable/LazyList.scala
@@ -66,5 +66,5 @@ object LazyList extends IterableFactory[LazyList] {
 
   def newBuilder[A]: Builder[A, LazyList[A]] = ???
 
-  def empty[A <: Any]: LazyList[A] = new LazyList[A](None)
+  def empty[A]: LazyList[A] = new LazyList[A](None)
 }

--- a/src/test/scala/strawman/collection/test/Test.scala
+++ b/src/test/scala/strawman/collection/test/Test.scala
@@ -248,6 +248,8 @@ class StrawmanTest {
     val ys15: Seq[(Int, Boolean)] = xs15
     val xs16 = xs.reverse
     val ys16: Seq[Int] = xs16
+    val ys17 = LazyList.empty ++ LazyList.empty
+    val ys18: LazyList[Nothing] = ys17
     println("-------")
     println(x1)
     println(x2)

